### PR TITLE
Dialogue: improve UX when creating new blocks

### PR DESCRIPTION
**Pull request by retrofox**
_This PR is part of the [follow-up tasks](https://github.com/Automattic/jetpack/issues/18123) after merging `conversation`/`dialogue` blocks._

This PR improves the UX when creating new `dialogue block.
Fixes #

#### Changes proposed in this Pull Request:

* Dialogue: improve UX when creating new blocks.

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Create a new `conversation` block
* Edit the participant names in order to make the testing easier to understand
* Confirm that it creates a new dialogue block when pressing ENTER, where
  * it picks up the next participant from the conversation block. If it's the last one, it will pick the first in the list.
  * If it breaks the block in the middle, the text os properly applied to both blocks 
  * It copies the timestamp of the selected block
* It behaves similarly when the new diaogue block is inserted and the previous block is a dialogue too.

https://user-images.githubusercontent.com/77539/102926887-97497180-4474-11eb-9474-c04e3951d00d.mov

#### Proposed changelog entry for your changes:

* Dialogue: improve UX when creating new blocks.

### CODE REVIEW

File: extensions/blocks/conversation/edit.js [hunk 1]
```diff
@@ -67,8 +62,24 @@ function ConversationEdit ( {
 
 	// Context bridge.
 	const contextProvision = {
-		setAttributes: useMemo( () => setAttributes, [ setAttributes ] ),
+		setAttributes,
 		updateParticipants,
+		getParticipantIndex: ( slug ) =>
+			participants.map( ( part ) => part.participantSlug ).indexOf( slug ),
+		getNextParticpantIndex: ( slug, offset = 0 ) => {
```
Comment (mreishus) [inline extensions/blocks/conversation/edit.js:69]: Small spelling error, missing an `i`. Recommend: `getNextParticpantIndex` -> `getNextParticipantIndex`

Comment (retrofox) [reply extensions/blocks/conversation/edit.js:69]: 🤦 Thanks, Matthew.

**Comment by mreishus** [comment]
Suggested refactor using modulus:
```js
getNextParticipantIndex: ( slug, offset = 0 ) => (contextProvision.getParticipantIndex( slug ) + 1 + offset) % participants.length;
```
Test code that shows it's mostly the same (runnable in F12 of new browser tab):
```js
let participants = [ 'alice', 'bob', 'charlie', 'dave' ];
let getParticipantIndex = slug => participants.indexOf( slug );
let getNextParticipantIndex1 = ( slug, offset = 0 ) => {
	let index = getParticipantIndex( slug ) + offset;
	if ( index < 0 ) {
		return 0;
	}

	if ( index + 1 >= participants.length ) {
		return 0;
	}

	return index + 1;
};
let getNextParticipantIndex2 = ( slug, offset = 0 ) => 
    (getParticipantIndex( slug ) + 1 + offset) % participants.length;

for (let p of participants.concat(["Does_not_exist"])) {
    for (let i = 0; i < 10; i++) {
        let newIndex1 = getNextParticipantIndex1(p, i);
        let newIndex2 = getNextParticipantIndex2(p, i);
        console.log({i, p, newIndex1, newIndex2});
    }
}

```
It does continue to "wrap" with larger offsets, so that may or may not be want you want. I believe it's easier to understand for people familiar w/ the modulus idiom.

**Comment by mreishus** [comment]
:heavy_check_mark: Pressing enter does create a new block that advances to the next participant.
:grey_question: I did notice a bug that might not be related to these changes.  When I add a new participant, his name isn't bolded like the others are.  See video:

https://user-images.githubusercontent.com/937354/103795216-51abad80-500b-11eb-8014-6cd5ac02b023.mp4

File: extensions/blocks/dialogue/edit.js [hunk 2]
```diff
@@ -84,11 +92,28 @@ export default function DialogueEdit ( {
 	const currentParticipant = getParticipantBySlug( participants, currentParticipantSlug );
 	const participantLabel = isCustomParticipant ? participant : currentParticipant?.participant;
 
-	const showTimestamp = isCustomParticipant ? showTimestampLocally : showTimestampGlobally;
-
 	// Conversation context. A bridge between dialogue and conversation blocks.
 	const transcritionBridge = useContext( ConversationContext );
 
+	// Set initial attributes according to the context.
```
Comment (mreishus) [inline extensions/blocks/dialogue/edit.js:98]: Another small spelling error (`transcritionBridge` -> `transcriptionBridge`)

**Comment by retrofox** [comment]
> ✔️ Pressing enter does create a new block that advances to the next participant.
> ❔ I did notice a bug that might not be related to these changes. When I add a new participant, his name isn't bolded like the others are. See video:

Nice catch. Something to fix. Probably we'll find more issues like this one 😅 . Just FYI, the new participant should get bold style activated as default.

File: extensions/blocks/dialogue/edit.js [hunk 3]
```diff
@@ -259,16 +284,38 @@ export default function DialogueEdit ( {
...
-						return createBlock( blockName );
-					}
-
 					return createBlock( blockName, {
 						...attributes,
 						content: value,
 					} );
 				} }
-				onReplace={ onReplace }
+				onReplace={ ( blocks, ...args ) => {
```
Comment (mreishus) [inline extensions/blocks/dialogue/edit.js:292]: Recommend a small comment at the top of this explaining its purpose; if I came back in 6 months and looked at it I might be confused.

**Review by mreishus — APPROVED** [review]
:+1: LGTM. Suggested optional refactors above.

**Comment by retrofox** [comment]
> Suggested refactor using modulus:

Thanks. Works like a charm. 👍

**Comment by mreishus** [comment]
Noticed one additional issue that is probably not related to the changes here, but only because of the recent rebase.

Earlier today, I could type what Alice says, then press enter, then type what Bob says.  Now, I need to type what Alice says, then press enter and press down arrow, then type what Bob says.

Video: 

https://user-images.githubusercontent.com/937354/103824620-e9c08b80-5039-11eb-9ad6-dd6413d54f34.mp4

**Review by mreishus — APPROVED** [review]
Retested and everything looks good, including timestamps.  Thanks for the comments! Those should help for anyone looking later. :heavy_check_mark: 

To recap, I did notice two issues with the conversation block that weren't introduced by these changes:
* Adding a Dave participant and his name isn't bolded
* Need to press enter, then down arrow to keep conversation flowing

**Review by jeherve — APPROVED** [review]

**Comment by retrofox** [comment]
> Retested and everything looks good, including timestamps. Thanks for the comments! Those should help for anyone looking later. ✔️

Thanks 🙇‍♂️ 

> To recap, I did notice two issues with the conversation block that weren't introduced by these changes:
> 
> * Adding a Dave participant and his name isn't bolded

https://github.com/Automattic/jetpack/issues/18222

> * Need to press enter, then down arrow to keep conversation flowing

https://github.com/Automattic/jetpack/issues/18228

deployed here r219065-wpcom

**Pull request merged**